horace_core/utilities/split_vector.m

```diff
@@ -0,0 +1,100 @@
+function [chunks, idxs] = split_vector(vector, max_chunk_sum)
+%SPLIT_VECTOR Split the given vector into a set of sub-vectors such that the
+% sum of each sub-vector has a maximum of max_chunk_sum, or the sub-vector has length
+% 1.
+%
+% If a value in `vector` is greater than max_chunk_sum, then that value will comprise
+% its own sub-vector.
+%
+% Input:
+% ------
+% vector       A vector of non-negative values.
+% max_chunk_sum      A positive value specifying the maximum sum for each sub-vector.
+%
+% Output:
+% -------
+% chunks       Cell array of vectors. Concatenation of these vectors will be
+%              equal to the input vector.
+% idxs         The indices at which the input vector was "split". Has size
+%              [2, n], where n is numel(chunks). Each idxs(:, i) is the
+%              lower and upper index into 'vector' of chunks{i}.
+%
+% Example:
+% --------
+% >> vector = [3, 2, 0, 6, 0, 5, 3, 1, 1, 24, 4, 2, 3, 0];
+% >> max_chunk_sum = 11;
+% >> [chunks, idxs] = split_vector(vector, max_chunk_sum)
+%   chunks =
+%       { [3, 2, 0, 6, 0], [5, 3, 1, 1], [24], [4, 2, 3, 0] }
+%   idxs =
+%       1     6    10    11
+%       5     9    10    14
+%
+if isempty(vector)
+    chunks = {};
+    idxs = zeros(2, 0);
+    return;
+end
+
+validateattributes(vector, {'numeric'}, {'vector', 'nonnegative'});
+validateattributes(max_chunk_sum, {'numeric'}, {'scalar', 'positive'});
+
+% We need the number of chunks we're to return so we can allocate output arrays.
+% Calculating the real number of chunks required is non-trivial. We can get the
+% minimum number using `sum(vector)/max_chunk_sum`, but this will result in us
+% under-allocating the output arrays.
+% If we under-estimate the number of chunks, we will end up re-allocating the
+% output arrays as they grow, so we'll need two instances of each array. Hence
+% it's more memory efficient to over-allocate to the maximum number of chunks
+% and then crop.
+max_num_chunks = numel(vector);
```

Comment:
I would extend this to say "... so initialize assuming all chunks will be contain one index" or something to explain why `numel(vector)` is the right value